tonic/src/service/interceptor.rs

```diff
@@ -35,24 +38,42 @@ use tower_service::Service;
 /// [tower]: https://crates.io/crates/tower
 /// [example]: https://github.com/hyperium/tonic/tree/master/examples/src/interceptor
 /// [tower-example]: https://github.com/hyperium/tonic/tree/master/examples/src/tower
-pub fn interceptor_fn<F>(f: F) -> InterceptorFn<F>
+pub trait Interceptor {
+    /// Intercept a request before it is sent, optionally cancelling it.
+    fn call(&mut self, request: crate::Request<()>) -> Result<crate::Request<()>, Status>;
+}
+
+impl<F> Interceptor for F
 where
     F: FnMut(crate::Request<()>) -> Result<crate::Request<()>, Status>,
+{
+    fn call(&mut self, request: crate::Request<()>) -> Result<crate::Request<()>, Status> {
+        self(request)
+    }
+}
+
+/// Create a new interceptor layer.
+///
+/// See [`Interceptor`] for more details.
+pub fn interceptor_fn<F>(f: F) -> InterceptorFn<F>
```

Comment:
I don't think this name is great, but of course we have to keep it for backwards compatibility. 

However, what if we did something like this:

```rust
#[deprecated(
    since = "0.5.1",
    note = "Please use the `interceptor` function instead"
)]
pub fn interceptor_fn<F>(f: F) -> InterceptorFn<F>
where
    F: Interceptor,
{
    Interceptor { inner: f }
}

#[deprecated(
    since = "0.5.1",
    note = "Please use the `Interceptor` type instead"
)]
pub type InterceptorFn<F> = Interceptor<F>;

pub struct Interceptor<T> {
    inner: T,
}

pub fn interceptor<T>(inner: T) -> Interceptor<T>
where
    T: Interceptor,
{
    Interceptor { inner }
}
```

Seems like the best of both worlds 🤔

Regarding the name of `fn interceptor` I personally prefer just calling it `interceptor` rather than `interceptor_layer` because that requires users knowing what a "layer" is. I don't feel very strongly about this though.

Reply:
I agree with you on all points,
but the name of the `Interceptor` struct is still an open question.

Maybe calling the struct `InterceptorLayer` isn't so bad, because the downstream user doesn't necessarily have to correlate the word "layer" to the `tower` layer, but it's still a "layer" in the abstract sense of the word. WDYT?

Reply:
And the function should still be named `interceptor`, I'm suggesting adding the word `Layer` just to the resulting struct's name.

Reply:
Yeah I think that's fine 👍